src/ctap/mod.rs

```diff
@@ -283,7 +283,7 @@ fn send_keepalive_up_needed(
             Ok(SendOrRecvStatus::Sent) => {
                 debug_ctap!(env, "Sent KEEPALIVE packet");
             }
-            Ok(SendOrRecvStatus::Received) => {
+            Ok(SendOrRecvStatus::Received(_)) => {
```

Comment:
What if we receive on a different endpoint than the one being kept alive? Does it mean we discard it?

Reply:
Good find.

At present we discard it. I've added a TODO to address this (for a later PR), but I can address it here is you'd prefer.

Reply:
I guess it's not really a regression (but I might be wrong). So I would say it's ok to do it later.

Reply:
I think it is a regression if we receive, but the channel ID is by chance the same. Then a different transport would interact with this channel.

Reply:
Yes, good point.
I can add a check for this.

However, I am wondering about a bigger change. If the code ignores packets on the other interface, then those packets are lost. So perhaps the better approach is to send() but not send_and_recv() at this time. That way the incoming packets are not serviced and will sit there. Looking at the code, it looks challenging to connect incoming packets to the other interface without massive refactoring.
Thoughts?

Reply:
I generally agree that a send / receive interface would be cleaner. The current solution is there for practical reasons. USB Request Blocks (URB) that need to be answered quickly. And if you only send and have the syscall roundtrip delay from that, you can't answer OUT packets in time.

If you try to go for a send / receive interface, please test for everything that didn't work last time.
@gendx @jmichelp are the original authors who made that decision. Maybe they can help describe a few test cases to look out for?

Reply:
OK.

What I've done in the short term is ignore those packet on the other interface. This stops the regression and leaves room for the above discussed improvements at a later date.
I've add tests to ./tools/vendor_hid_tests.py to verify this behavior.